### PR TITLE
crimson/os/seastore: minimize the static size of onode_layout_t 

### PR DESCRIPTION
**Pull request by cyx1231st**
This reduces the onode-tree write amplification (MUTATE transaction) from ~1.6 to ~0.3, which is 5+ times better:
Before:
before-1
After:
after-1

However, the test doesn't write other types of attrs. We still need to implement flexible-sized onode value for better performance in more common scenarios.

Apart from that, the major write amplification now comes from inefficient cleaning and high transaction conflict rate, see:
after-2
invalidate-ratio

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by cyx1231st** [comment]
BTW, still hitting pinset issue after 10 minuites write:
```
57087 INFO  2021-09-10 16:15:07,678 [shard 0] seastore - OTree::LeafNode::insert_value(0x60000016e1c0): proceed split NodeL0@0x1e1bddcf4000000+4000Lv0 to fresh NodeL0@0x1e141ad7e000000+4000Lv0 ...
57088 INFO  2021-09-10 16:15:07,678 [shard 0] seastore - OTree::Layout::split_insert: begin at insert_pos(36, 0, 0), insert_stage=2, insert_size=294B ...
57089 INFO  2021-09-10 16:15:07,678 [shard 0] seastore - OTree::Layout::split_insert: done  at insert_pos(8, 0, 0), insert_stage=2, insert_size=294B, split_at(28, 0, 0), is_insert_left=false, split_size=8219B(target=8216B)
57090 ERROR 2021-09-10 16:15:07,694 [shard 0] seastore - add_pin: unable to add btree_range_pin_t(begin=0, end=18446744073709551615, depth=3, extent=0x6000073a5380) (CachedExtent(addr=0x6000073a5380, type=LADDR_INTERNAL, version=0, dirty_from_or_retired_at=journal_seq_t(segment_seq=0, offset=paddr_t<NULL_SEG, NULL_OFF>), paddr=paddr_t<60, 32768>, state=CLEAN, last_committed_crc=3810898430, refcount=6, size=24, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=3))), found btree_range_pin_t(begin=0, end=18446744073709551615, depth=3, extent=0x60000223ad80) (CachedExtent(addr=0x60000223ad80, type=LADDR_INTERNAL, version=16, dirty_from_or_retired_at=journal_seq_t(segment_seq=4294967295, offset=paddr_t<4294967295, NULL_OFF>), paddr=paddr_t<60, 32768>, state=INVALID, last_committed_crc=1190232079, refcount=2, size=25, meta=btree_node_meta_t(begin=0, end=18446744073709551615, depth=3)))
57091 ERROR 2021-09-10 16:15:07,694 [shard 0] none - ../src/crimson/os/seastore/lba_manager/btree/btree_range_pin.cc:122 : In function 'void crimson::os::seastore::lba_manager::btree::btree_pin_set_t::add_pin(crimson::os::seastore::lba_manager::btree::btree_range_pin_t&)', ceph_assert(%s)
57092 0 == "impossible"
```

**Review by liu-chunmei — APPROVED** [review]

File: src/osd/PGLog.cc [hunk 1]
```diff
@@ -996,7 +996,12 @@ void PGLog::rebuild_missing_set_with_deletes(
     dout(20) << __func__ << " check for log entry: " << *i << " = " << r << dendl;
 
     if (r >= 0) {
+#ifndef WITH_SEASTAR
       object_info_t oi(bv);
+#else
+      object_info_t oi;
+      oi.decode_without_oid(bv);
+#endif
```
Comment (athanatos) [inline src/osd/PGLog.cc:1004]: This pathway can't happen for crimson at all at the moment.

**Comment by athanatos** [comment]
I don't want to hack up the object_info_t encode and decode pathways.  Right now, classic and crimson share encoding formats for identical types.  If we use compile directives like this to break that symmetry, it will make compatibility later on a nightmare.

File: src/osd/PGLog.h [hunk 2]
```diff
@@ -1532,7 +1532,12 @@ struct PGLog : DoutPrefixProvider {
 	    OI_ATTR,
 	    bv);
 	  if (r >= 0) {
+#ifndef WITH_SEASTAR
 	    object_info_t oi(bv);
+#else
+	    object_info_t oi;
+	    oi.decode_without_oid(bv);
+#endif
```
Comment (athanatos) [inline src/osd/PGLog.h:1540]: This pathway is also exclusive to classic.

Comment (athanatos) [reply src/osd/PGLog.h:1540]: The crimson version is read_log_and_missing_crimson.

File: src/osd/osd_types.cc [hunk 3]
```diff
@@ -6178,24 +6178,30 @@ void object_info_t::copy_user_bits(const object_info_t& other)
 
 void object_info_t::encode(ceph::buffer::list& bl, uint64_t features) const
```
Comment (athanatos) [inline src/osd/osd_types.cc:6179]: There are places where we encode object_info_t other than the '_' attr on the object.  Hacking up the encode/decode method here will break those users.  Can we modify crimson to write a different, truncated object_info_t (object_info_no_oid_t) in the attr?  We actually want to do this in classic as well.

**Comment by athanatos** [comment]
I don't want to do this hack.  I'd rather invest the time required to properly refactor crimson and/or classic to update just encoding stored on '_'.

**Comment by cyx1231st** [comment]
> Messing with object_info_t encode/decode universally is going to create problems. I'd rather invest the time required to properly refactor crimson and/or classic to update just the encoding stored on '_'.

I think so, I wanted to leverate the first version to collect feedback by marking out all the potential users of `object_info_t` denc. I think I've got the clues about how to do next.

File: src/osd/PGLog.cc (hunk 1 above)
Comment (cyx1231st) [reply src/osd/PGLog.cc]: dropped

File: src/osd/PGLog.h (hunk 2 above)
Comment (cyx1231st) [reply src/osd/PGLog.h]: dropped

File: src/osd/osd_types.cc (hunk 3 above)
Comment (cyx1231st) [reply src/osd/osd_types.cc:6179]: Added the separate `encode_no_oid()` and `decode_no_oid()` for crimson path. Also use `ceph_abort()` to make sure we won't hit the original denc path in crimson.

File: src/osd/osd_types.cc [hunk 4]
```diff
@@ -6312,6 +6318,59 @@ void object_info_t::decode(ceph::buffer::list::const_iterator& bl)
   DECODE_FINISH(bl);
 }
```
Comment (athanatos) [inline src/osd/osd_types.cc:6320]: I'd rather we clear the oid field and use the existing encode/decode calls.  Doing it this way means that we'll crash when decoding an object_info_t encoded by classic.

Medium term, we probably want to simply remove the oid field from both classic and crimson and update the encoding version.

Comment (athanatos) [reply src/osd/osd_types.cc:6320]: That is, encode_no_oid would swap out the oid field, encode, and then restore it.

Comment (cyx1231st) [reply src/osd/osd_types.cc]: Fixed.

**Review by athanatos — APPROVED** [review]

**Pull request merged**